test/integration/jobschedulingpreference_test.go

```diff
@@ -0,0 +1,245 @@
+
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	fedschedulingv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/scheduling/v1alpha1"
+	clientset "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/pkg/schedulingtypes"
+	"github.com/kubernetes-sigs/federation-v2/pkg/schedulingtypes/adapters"
+	"github.com/kubernetes-sigs/federation-v2/test/common"
+	"github.com/kubernetes-sigs/federation-v2/test/integration/framework"
+	batchv1 "k8s.io/api/batch/v1"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	restclient "k8s.io/client-go/rest"
+)
+
+// TestReplicaSchedulingPreference validates basic replica scheduling preference calculations.
+var TestJobSchedulingPreference = func(t *testing.T) {
+	t.Parallel()
+	tl := framework.NewIntegrationLogger(t)
+
+	controllerFixture, fedClient := initJSPTest(tl, FedFixture)
+	defer controllerFixture.TearDown(tl)
+
+	clusters := FedFixture.ClusterNames()
+	if len(clusters) != 2 {
+		tl.Fatalf("Expected two clusters to be part of Federation Fixture setup")
+	}
+
+	kubeClient := FedFixture.KubeApi.NewClient(tl, "jsp")
+	testNamespace := framework.CreateTestNamespace(tl, kubeClient, "jsp-")
+
+	testCases := map[string]struct {
+		jspSpec  fedschedulingv1a1.JobSchedulingPreferenceSpec
+		expected map[string]adapters.JobSchedulingResult
+	}{
+		"Distribution happens evenly across all clusters": {
+			jspSpec: jspSpecWithoutClusterWeights(int32(4), int32(4)),
+			expected: map[string]adapters.JobSchedulingResult{
+				clusters[0]: {
+					Parallelism: int32(2),
+					Completions: int32(2),
+				},
+				clusters[1]: {
+					Parallelism: int32(2),
+					Completions: int32(2),
+				},
+			},
+		},
+	}
+
+	for testName, testCase := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			name, err := createJSPTestObjs(testCase.jspSpec, testNamespace, fedClient)
+			if err != nil {
+				tl.Fatalf("Creation of test objects failed in federation")
+			}
+
+			err = waitForMatchingJobPlacement(fedClient, name, testNamespace, testCase.expected)
+			if err != nil {
+				tl.Fatalf("Failed waiting for matching placements")
+			}
+
+			err = waitForMatchingJobOverride(fedClient, name, testNamespace, testCase.expected)
+			if err != nil {
+				tl.Fatalf("Failed waiting for matching overrides")
+			}
+		})
+	}
+}
+
+func initJSPTest(tl common.TestLogger, fedFixture *framework.FederationFixture) (*framework.ControllerFixture, clientset.Interface) {
+	config := fedFixture.KubeApi.NewConfig(tl)
+	restclient.AddUserAgent(config, "rsp-test")
+	kubeClient := fedFixture.KubeApi.NewClient(tl, "rsp-test")
+	fedClient := clientset.NewForConfigOrDie(config)
+	err := common.EnableType(kubeClient, fedClient, schedulingtypes.JobTypeName, util.DefaultFederationSystemNamespace)
```

Comment:
I don't think it should be necessary to create the type config in the API like this.  Consider passing the typeconfig into the scheduler constructor so that tests can pass the typeconfig directly and whatever starts the scheduler in the controller manager can be responsible for retrieving the appropriate type config from the API.